File: app/src/main/java/zmuzik/ubike/di/AppModule.java
package zmuzik.ubike.di;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.location.LocationManager;
import android.preference.PreferenceManager;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import zmuzik.ubike.App;

@Module
public class AppModule {

    Application mApplication;

    public AppModule(Application application) {
        mApplication = application;
    }

    @Provides
    @AppContext
    Context provideApplicationContext() {
        return mApplication.getApplicationContext();
    }

    @Provides
    Application provideApplication() {
        return mApplication;
    }

    @Provides
    App provideApp() {
        return (App) mApplication;
    }

    @Provides
    SharedPreferences provideSharedPreferences(Application application) {
        return PreferenceManager.getDefaultSharedPreferences(application);
    }

    @Provides
    LocationManager provideLocationManager() {
        return (LocationManager) mApplication.getSystemService(Context.LOCATION_SERVICE);
    }

    @Provides
    OkHttpClient provideOkHttp() {
        return new OkHttpClient();
    }

}

File: app/src/main/java/zmuzik/ubike/di/AppModule.kt
package zmuzik.ubike.di

import android.app.Application
import android.content.Context
import android.location.LocationManager
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import zmuzik.ubike.Presenter
import javax.inject.Singleton

@Module
class AppModule(private val application: Application) {

    @Provides
    @Singleton
    fun provideApplicationContext(): Context {
        return application
    }

    @Provides
    @Singleton
    fun provideLocationManager(): LocationManager {
        return application.getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }

    @Provides
    @Singleton
    fun provideHttpClient(): OkHttpClient {
        return OkHttpClient()
    }

    @Provides
    @Singleton
    fun providePresenter(): Presenter {
        return Presenter()
    }
}